django/forms/widgets.py

```diff
@@ -48,15 +48,32 @@ def __init__(self, media=None, css=None, js=None):
                 css = {}
             if js is None:
                 js = []
-        self._css = css
-        self._js = js
+        self._css_lists = [css]
+        self._js_lists = [js]
 
     def __repr__(self):
         return 'Media(css=%r, js=%r)' % (self._css, self._js)
 
     def __str__(self):
         return self.render()
 
+    @property
+    def _css(self):
+        css = self._css_lists[0]
+        for obj in filter(None, self._css_lists[1:]):
+            css = {
+                medium: self.merge(css.get(medium, []), obj.get(medium, []))
+                for medium in css.keys() | obj.keys()
+            }
+        return css
+
+    @property
+    def _js(self):
+        js = self._js_lists[0]
+        for obj in filter(None, self._js_lists[1:]):
+            js = self.merge(js, obj)
+        return js
```

Comment:
Can this be rewritten as

```python
from functools import reduce
return reduce(self.merge, filter(None, self._js_lists))
```

By the way, how can entries be `None` given the checks in `__init__`? If it's not necessay than this can be reduced to

```python
return reduce(self.merge, self._js_lists)
```

Reply:
Thanks! Using `reduce` should be possible, yes. That's a good idea. I assumed that you have to provide an initializer (similar to `sum`) but it seems `reduce` is smarter here.

The `filter` isn't strictly necessary but it avoids calling `self.merge` for empty lists and dicts:

```
>>> lists = [[1], [], [2]]
>>> list(filter(None, lists))
[[1], [2]]
```

Reply:
The list may be empty after the filter call if there are no JS assets in there. Also, the return value should be a list again, and the tests even assume a tuple in some places.

If it's OK with you I'd rather leave the current code in there instead of `return list(reduce(self.merge, filter(None, self._js_lists), []))` and changing unrelated tests. I'm less fond of reduce/map/lambda than I used to be...